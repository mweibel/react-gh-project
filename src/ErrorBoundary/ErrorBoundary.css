.error-details {
  white-space: pre-wrap;
}