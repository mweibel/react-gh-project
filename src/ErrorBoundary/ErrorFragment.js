import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

const ErrorFragment = ({ error, errorInfo = null }) => (
  <Fragment>
    <h2>Oops</h2>
    <p>An error happened.</p>
    <details className="error-details">
      {error && error.toString()}
      <br />
      {errorInfo !== null && errorInfo.componentStack}
    </details>
  </Fragment>
)
ErrorFragment.propTypes = {
  error: PropTypes.shape({
    toString: PropTypes.func.isRequired
  }),
  errorInfo: PropTypes.shape({
    componentStack: PropTypes.string.isRequired
  })
}
export default ErrorFragment;