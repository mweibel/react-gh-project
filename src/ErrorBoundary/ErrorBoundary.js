import React, { Component } from 'react';
import ErrorFragment from './ErrorFragment';
import './ErrorBoundary.css';

class ErrorBoundary extends Component {
  constructor(props) {
    super(props);
    this.state = {
      error: null,
      errorInfo: null
    };
  }

  componentDidCatch(error, errorInfo) {
    this.setState({
      error: error,
      errorInfo: errorInfo
    });
  }

  render() {
    const { error, errorInfo } = this.state;

    if (error || errorInfo) {
      return <ErrorFragment error={error} errorInfo={errorInfo} />;
    }
    return this.props.children;
  }
}

export default ErrorBoundary;
