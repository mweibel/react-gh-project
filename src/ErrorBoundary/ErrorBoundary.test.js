import React from 'react';
import ErrorBoundary from './ErrorBoundary';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme'

it('renders children if no error happened', () => {
  const tree = renderer.create(
    <ErrorBoundary>
      <div>a child component</div>
    </ErrorBoundary>
  );
  expect(tree).toMatchSnapshot();
});

it('renders an error if an error happens', () => {
  expect.assertions(3);

  // ErrorBoundary logs an error also in tests, suppress those
  // so devs aren't confused when seeing such an error but the test itself passing.
  // See: https://github.com/facebook/react/issues/11098 -- as of react 16.4.3+ this
  // is avoidable, but that version is not out yet.
  spyOn(console, 'error');

  const ErroringComponent = () => {
    throw new Error('an Error')
  }
  const wrapper = mount(
    <ErrorBoundary>
      <ErroringComponent />
    </ErrorBoundary>
  );

  expect(wrapper.state('error')).not.toBe(null)
  expect(wrapper.state('errorInfo')).not.toBe(null)
  expect(wrapper.find('details').length).toBe(1);
})