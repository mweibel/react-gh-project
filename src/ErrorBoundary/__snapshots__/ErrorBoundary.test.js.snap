// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders children if no error happened 1`] = `
<div>
  a child component
</div>
`;
